parachain/test-parachains/adder/collator/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Collator for the adder test parachain.
+
+use std::{pin::Pin, sync::{Arc, Mutex}};
+use test_parachain_adder::{hash_state, BlockData, HeadData, execute};
+use futures::{Future, FutureExt};
+use polkadot_primitives::v1::{ValidationData, PoV, Hash};
+use polkadot_node_primitives::Collation;
+use codec::Encode;
+
+/// The amount we add when producing a new block.
+///
+/// This is a constant to make tests easily reproducible.
+const ADD: u64 = 2;
+
+/// The state of the adder parachain.
+struct State {
+	genesis_state: HeadData,
+	last_head: HeadData,
+	state: u64,
+}
+
+impl State {
+	/// Init the genesis state.
+	fn genesis() -> Self {
+		let genesis_state = HeadData {
+			number: 0,
+			parent_hash: Default::default(),
+			post_state: hash_state(0),
+		};
+		let last_head = genesis_state.clone();
+
+		Self {
+			genesis_state,
+			last_head,
+			state: 0,
+		}
+	}
+
+	/// Advance the state and produce a new block.
+	///
+	/// Returns the new [`BlockData`] and the new [`HeadData`].
+	fn advance(&mut self) -> (BlockData, HeadData) {
+		let block = BlockData {
+			state: self.state,
+			add: ADD,
+		};
+		let new_head = execute(self.last_head.hash(), self.last_head.clone(), &block)
+			.expect("Produces valid block");
+
+		self.last_head = new_head.clone();
+		self.state = self.state.wrapping_add(ADD);
+
+		(block, new_head)
+	}
+}
+
+/// The collator of the adder parachain.
+pub struct Collator {
+	state: Arc<Mutex<State>>,
+}
+
+impl Collator {
+	/// Create a new
```

Comment:
seems to be unfinished.

We could write like:

```suggestion
	/// Create a new collator instance with the state initialized as genesis.
```